File: swift/SDTrustKit/Sources/SDTrustKit/Options.swift
import Foundation

public struct VerificationOptions: Encodable, Equatable, Sendable {
    public var signerTrustAnchorsDerBase64: [String]
    public var signerTrustAnchorSets: [TimedTrustAnchorSet]
    public var timestampTrustAnchorsDerBase64: [String]
    public var timestampTrustAnchorSets: [TimedTrustAnchorSet]
    public var timestampCertificateSha256Pins: [String]

    public init(
        signerTrustAnchorsDer: [Data] = [],
        signerTrustAnchorSets: [TimedTrustAnchorSet] = [],
        timestampTrustAnchorsDer: [Data] = [],
        timestampTrustAnchorSets: [TimedTrustAnchorSet] = [],
        timestampCertificateSha256Pins: [String] = []
    ) {
        self.signerTrustAnchorsDerBase64 = signerTrustAnchorsDer.map { $0.base64EncodedString() }
        self.signerTrustAnchorSets = signerTrustAnchorSets
        self.timestampTrustAnchorsDerBase64 = timestampTrustAnchorsDer.map { $0.base64EncodedString() }
        self.timestampTrustAnchorSets = timestampTrustAnchorSets
        self.timestampCertificateSha256Pins = timestampCertificateSha256Pins
    }
}

public struct TimedTrustAnchorSet: Encodable, Equatable, Sendable {
    public var validFromUnixSeconds: Double?
    public var validUntilUnixSeconds: Double?
    public var anchorsDerBase64: [String]

    public init(
        validFromUnixSeconds: Double? = nil,
        validUntilUnixSeconds: Double? = nil,
        anchorsDer: [Data]
    ) {
        self.validFromUnixSeconds = validFromUnixSeconds
        self.validUntilUnixSeconds = validUntilUnixSeconds
        self.anchorsDerBase64 = anchorsDer.map { $0.base64EncodedString() }
    }
}

public struct RevocationOptions: Encodable, Equatable, Sendable {
    public var nowUnixSeconds: Double?
    public var crlCacheEntries: [CrlCacheEntry]

    public init(nowUnixSeconds: Double? = nil, crlCacheEntries: [CrlCacheEntry] = []) {
        self.nowUnixSeconds = nowUnixSeconds
        self.crlCacheEntries = crlCacheEntries
    }
}

public struct CrlCacheEntry: Encodable, Equatable, Sendable {
    public var url: String?
    public var cacheKeySha256: String?
    public var validUntilUnixSeconds: Double
    public var derBase64: String

    public init(
        url: String,
        validUntilUnixSeconds: Double,
        der: Data
    ) {
        self.url = url
        self.cacheKeySha256 = nil
        self.validUntilUnixSeconds = validUntilUnixSeconds
        self.derBase64 = der.base64EncodedString()
    }

    public init(
        cacheKeySha256: String,
        validUntilUnixSeconds: Double,
        der: Data
    ) {
        self.url = nil
        self.cacheKeySha256 = cacheKeySha256
        self.validUntilUnixSeconds = validUntilUnixSeconds
        self.derBase64 = der.base64EncodedString()
    }
}

extension JSONEncoder {
    static var rustOptions: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }
}

File: kotlin/SDTrustKit/sdtrustkit/src/main/java/com/sdtrustkit/Options.kt
package com.sdtrustkit

import android.util.Base64
import kotlinx.serialization.Serializable

@Serializable
data class VerificationOptions(
    val signerTrustAnchorsDerBase64: List<String> = emptyList(),
    val signerTrustAnchorSets: List<TimedTrustAnchorSet> = emptyList(),
    val timestampTrustAnchorsDerBase64: List<String> = emptyList(),
    val timestampTrustAnchorSets: List<TimedTrustAnchorSet> = emptyList(),
    val timestampCertificateSha256Pins: List<String> = emptyList(),
) {
    companion object {
        fun fromDer(
            signerTrustAnchorsDer: List<ByteArray> = emptyList(),
            signerTrustAnchorSets: List<TimedTrustAnchorSet> = emptyList(),
            timestampTrustAnchorsDer: List<ByteArray> = emptyList(),
            timestampTrustAnchorSets: List<TimedTrustAnchorSet> = emptyList(),
            timestampCertificateSha256Pins: List<String> = emptyList(),
        ): VerificationOptions =
            VerificationOptions(
                signerTrustAnchorsDerBase64 = signerTrustAnchorsDer.map(::base64),
                signerTrustAnchorSets = signerTrustAnchorSets,
                timestampTrustAnchorsDerBase64 = timestampTrustAnchorsDer.map(::base64),
                timestampTrustAnchorSets = timestampTrustAnchorSets,
                timestampCertificateSha256Pins = timestampCertificateSha256Pins,
            )
    }
}

@Serializable
data class TimedTrustAnchorSet(
    val validFromUnixSeconds: Double? = null,
    val validUntilUnixSeconds: Double? = null,
    val anchorsDerBase64: List<String> = emptyList(),
) {
    companion object {
        fun fromDer(
            validFromUnixSeconds: Double? = null,
            validUntilUnixSeconds: Double? = null,
            anchorsDer: List<ByteArray>,
        ): TimedTrustAnchorSet =
            TimedTrustAnchorSet(
                validFromUnixSeconds = validFromUnixSeconds,
                validUntilUnixSeconds = validUntilUnixSeconds,
                anchorsDerBase64 = anchorsDer.map(::base64),
            )
    }
}

@Serializable
data class RevocationOptions(
    val nowUnixSeconds: Double? = null,
    val crlCacheEntries: List<CrlCacheEntry> = emptyList(),
)

@Serializable
data class CrlCacheEntry(
    val url: String? = null,
    val cacheKeySha256: String? = null,
    val validUntilUnixSeconds: Double,
    val derBase64: String,
) {
    companion object {
        fun fromUrl(
            url: String,
            validUntilUnixSeconds: Double,
            der: ByteArray,
        ): CrlCacheEntry =
            CrlCacheEntry(
                url = url,
                validUntilUnixSeconds = validUntilUnixSeconds,
                derBase64 = base64(der),
            )

        fun fromCacheKey(
            cacheKeySha256: String,
            validUntilUnixSeconds: Double,
            der: ByteArray,
        ): CrlCacheEntry =
            CrlCacheEntry(
                cacheKeySha256 = cacheKeySha256,
                validUntilUnixSeconds = validUntilUnixSeconds,
                derBase64 = base64(der),
            )
    }
}

private fun base64(bytes: ByteArray): String =
    Base64.encodeToString(bytes, Base64.NO_WRAP)
